src/Features/Core/Portable/Snippets/AbstractSnippetProvider.cs

```diff
@@ -0,0 +1,134 @@
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImport;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.EditAndContinue.Contracts;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Snippets
+{
+    internal abstract class AbstractSnippetProvider : ISnippetProvider
+    {
+        public abstract string SnippetIdentifier { get; }
+        public abstract string SnippetDisplayName { get; }
+
+        protected readonly SyntaxAnnotation _cursorAnnotation = new();
+        protected readonly SyntaxAnnotation _findSnippetAnnotation = new();
+
+        /// <summary>
+        /// Uses the SyntaxContext to determine if the location the cursor is at is a valid place
+        /// for each corresponding SnippetProvider
+        /// </summary>
+        protected abstract Task<bool> IsValidSnippetLocationAsync(Document document, int position, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Generates the new snippet's TextChange's that are being inserted into the document
+        /// </summary>
+        protected abstract Task<ImmutableArray<TextChange>> GenerateSnippetTextChangesAsync(Document document, int position, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Method for each snippet to locate the inserted SyntaxNode to reformat
+        /// </summary>
+        protected abstract Task<SyntaxNode> AnnotateNodesToReformatAsync(Document document, SyntaxAnnotation reformatAnnotation, SyntaxAnnotation cursorAnnotation, int position, CancellationToken cancellationToken);
+        protected abstract int? GetTargetCaretPosition(ISyntaxFactsService syntaxFacts, SyntaxNode caretTarget);
+
+        /// <summary>
+        /// Determines if the location is valid for a snippet,
+        /// if so, then it creates a SnippetData.
+        /// </summary>
+        public async Task<SnippetData?> GetSnippetDataAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken))
+            {
+                return null;
+            }
+
+            if (await IsValidSnippetLocationAsync(document, position, cancellationToken).ConfigureAwait(false))
+            {
+                return new SnippetData(SnippetDisplayName, SnippetIdentifier);
+            }
+
+            return null;
```

Comment:
```suggestion
            if (!await IsValidSnippetLocationAsync(document, position, cancellationToken).ConfigureAwait(false))
            {
                return null;
            }

            return new SnippetData(SnippetDisplayName, SnippetIdentifier);
```

Reply:
general idea here being to be consistent with how you bail out or return values.  it makes tracking overal logic and data flow easier.